0000-scheduled-circuit.md

```diff
@@ -0,0 +1,266 @@
+# [WIP] [RFC] Scheduled Circuit
+
+| **Status**        | **Proposed/Accepted/Deprecated** |
+|:------------------|:---------------------------------------------|
+| **RFC #**         | ####                                         |
+| **Authors**       | Chris Wood (cjwood@us.ibm.com), Toshinari Itoko (itoko@jp.ibm.com), Ali Javadi (Ali.Javadi@ibm.com)    |
+| **Deprecates**    | NA                 |
+| **Submitted**     | 2020-04-DD                                   |
+| **Updated**       | YYYY-MM-DD                                   |
+
+## Summary
+Add "scheduled circuit" in-between “gate circuit”(QuantumCircuit) and “pulse schedule”(pulse.Schedule)
+to make time-aware circuit-based noise simulation possible.
+
+## Motivation
+Currently there is no way to instruct `delay` (or `wait`/`idle`) for a specific duration in circuit level.
+This prevents from easy noise simulation of experiments with `delay`s, e.g. T1/T2 measurement, on Aer.
+A new circuit class for incorporating with instructions having `duration` should be added,
+as well as scheduling methods to construct it from a QuantumCircuit.
+
+### Main use cases
+1) accurate simulation of decoherence in Aer (T1&T2, Cross-talk)
+2) insert gates in idle timeslots
+
+## User Benefit
+- Time-aware circuit-based noise simulation on Aer is enabled.
+- Scheduling in circuit level will enable more optimization than that in pulse level in the future.
+
+## Design Proposal
+### High-level API:
+1. Noise simulation flow (new):
+```
+timed_circuit = circuit.schedule(transpiled_circuit, ...)  # transpiled_circuit: QuantumCircuit
+qobj = assemble(timed_circuit)  # timed_circuit: ScheduledCircuit
+```
+
+2. Circuit-to-Schedule flow (update):
+```
+schedule = scheduler.circuit_schedule(transpiled_circuit, ...)  # transpiled_circuit: QuantumCircuit
+```
+->
+```
+timed_circuit = circuit.schedule(transpiled_circuit, ...)  # transpiled_circuit: QuantumCircuit
+schedule = pulse.build_schedule(timed_circuit, ...)  # timed_circuit: ScheduledCircuit
```

Comment:
```suggestion
schedule = pulse.lower_circuit(timed_circuit, ...)  # timed_circuit: ScheduledCircuit
```

How do you plan to establishing timing for circuits in the case where an instruction blocks resources it is not directly operating on at a semantic level? For example a cross talk compensation pulse inserted for a neighboring qubit. To do this properly likely requires the notion of both target (operated) and resource acquired qubits. If you were to schedule this at the circuit level and then lower to the pulse level (`build_schedule`) it would result in two different programs being executed with different timings. I think the desired behavior is for the timing to be equal.